### ISSUE TITLE
Bug: Printer fails to write generics for superclasses after setting it

### ISSUE THREAD

**Issue by MartinWitt** [INRIA/spoon#4057]
```java
    @Test
    void test() {
        CtType<?> type = Launcher.parseClass("class Foo<E> {}");
        CtType<?> foo = type.getReference().getTypeDeclaration();
        type.setSuperclass(foo.getReference());
        List<CtTypeParameter> typeParameters = type.getSuperclass().getTypeDeclaration().getFormalCtTypeParameters(); // has size 1
        String printedClass = type.toString(); // drops the generic type on the extended class. output: "class Foo<E> extends Foo {}"
    }
```
The example above shows the problem. The expected output is `class Foo<E> extends Foo<E> {}` but currently the printer drops the generic type. After some debugging I found something, that could be the problem:  
[ElementPrinterHelper](https://github.com/INRIA/spoon/blob/e8a1596aa29d8d34f84087fb0dbce1d032b2259f/src/main/java/spoon/reflect/visitor/ElementPrinterHelper.java#L250) `final Collection<CtTypeReference<?>> arguments = ctGenericElementReference.getActualTypeArguments();`returns an empty list. 
This looks like not a printer related problem, but more like a problem in setting a generic super class. Did I miss some method for setting a generic superclass explicitly with parameters? The type parameters for the super class are strangely correct in the model.

**Comment by slarse** [comment]
Hi @MartinWitt,

The output you get is correct given the code. What you're missing here to get your expected output is a type argument to the reference to that generic type. Note that when you get a reference to a generic type, it will by default not have any type arguments (because what would a "default" type argument be?). So I think that you may have confused the declared type parameter with an actual type argument.

To get your expected output, you'd need to do something like this:

```java
CtType<?> genericType = Launcher.parseClass("class Foo<E> {}");

CtTypeReference<?> genericTypeRef = type.getReference();
CtTypeParameter typeParam = type.getFormalCtTypeParameters().get(0);
genericTypeRef.addActualTypeArgument(typeParam.getReference());

type.setSuperclass(genericTypeRef);

System.out.println(type); // prints "class Foo<E> extends Foo<E> {}"
```

Note that of course, this is not valid Java code due to the cyclic reference, but I\m guessing that was not important for your example.

**Comment by slarse** [comment]
@MartinWitt did my suggestion resolve your problem?

**Comment by MartinWitt** [comment]
Yes, it works perfect. But it seems a bit unintuitive and could be documented somewhere?

**Comment by slarse** [comment]
> But it seems a bit unintuitive and could be documented somewhere?

Hmm, any suggestion of where to document it based on where you looked for a solution to your problem?

This is the same concept as creating a method invocation, you wouldn't expect it to have arguments unless you add them. So it's not hard to explain intuitively, at least.
